File: csharp/GildedRose.cs
using System.Collections.Generic;

namespace csharp
{
    public class GildedRose
    {
        IList<Item> Items;
        public GildedRose(IList<Item> items)
        {
            Items = items;
        }

        public void UpdateQuality()
        {

            foreach (var item in Items)
            {
                var qualityModifier = 1;
                var sellInModifier = 1;


                switch(item.Name)
                {
                    case "Sulfuras, Hand of Ragnaros":
                        qualityModifier = 0;
                        sellInModifier = 0;
                        break;
                    case "Aged Brie":
                        qualityModifier = -1;
                        break;
                    case "Conjured Mana Cake":
                        qualityModifier = 2;
                        break;
                    case "Backstage passes to a TAFKAL80ETC concert":
                        if (item.SellIn > 10)
                        {
                            qualityModifier = -1;
                        }
                        else if (item.SellIn > 5 && item.SellIn <= 10)
                        {
                            qualityModifier = -2;
                        }
                        else
                        {
                            qualityModifier = -3;
                        }
                        break;
                }

                if (item.SellIn <= 0)
                {
                    qualityModifier= qualityModifier * 2;
                }

                item.Quality = item.Quality - (1 * qualityModifier);

                if (item.Quality < 0)
                {
                    item.Quality = 0;
                }
                if (item.Quality > 50)
                {
                    item.Quality = 50;
                }

                item.SellIn = item.SellIn - (1 * sellInModifier);

                if (item.Name == "Backstage passes to a TAFKAL80ETC concert" && item.SellIn <= 0)
                {
                    item.Quality = 0;
                }


            }
            
        }
    }
}

File: TypeScript/app/gilded-rose.ts
export class Item {
    name: string;
    sellIn: number;
    quality: number;

    constructor(name, sellIn, quality) {
        this.name = name;
        this.sellIn = sellIn;
        this.quality = quality;
    }
}
class functionLookup{
    functionName;
    valueModifier;
    constructor(functionName, valueModifier){
        this.functionName = functionName;
        this.valueModifier = valueModifier;
    }
}
export class GildedRose {
    items: Array<Item>;

    constructor(items = []) {
        this.items = items;
    }

    updateQuality() {

        for (let item of this.items) {
            let identifier = GildedRose.findIdentifier(item.name);
            let functionMap = {
                "Sulfuras": new functionLookup(function() {},0),
                "Backstage pass": new functionLookup(backstagePass,0),
                "Aged Brie": new functionLookup(normalItem,-1),
                "Conjured": new functionLookup(normalItem,2),
                "Normal": new functionLookup(normalItem,1),};
            functionMap[identifier].functionName(item, functionMap[identifier].valueModifier);

        }
        return this.items;

        function backstagePass(item) {
            GildedRose.degradeItemQuality(item, GildedRose.backStagepassQualityIncrease(item))
        }
        function normalItem(item, value) {
            GildedRose.degradeItemQuality(item, value);
            GildedRose.ageItem(item);
        }

    }
    static degradeItemQuality(item, valueToDegradeQuality){
        if (item.sellIn <= 0){valueToDegradeQuality *= 2}
        item.quality -= valueToDegradeQuality;
        if (item.quality < 0){item.quality = 0}
        if (item.quality > 50){item.quality = 50}
        return item;
    }

    static ageItem(item){
        return item.sellIn--
    }

    static backStagepassQualityIncrease(item){
        if (item.sellIn > 10){return -1}
        if (item.sellIn > 5){return -2}
        if (item.sellIn > 0){return -3}
        else {return Infinity}
    }

    static findIdentifier(itemName){
        if (itemName.includes('Sulfuras')){return 'Sulfuras' }
        else if (itemName.includes('Conjured')){return 'Conjured' }
        else if (itemName.includes('Backstage pass')){return 'Backstage pass'}
        else if (itemName === 'Aged Brie'){return 'Aged Brie'}
        else {return 'Normal'}
    }
}